src/print.jl

```diff
@@ -1,9 +1,44 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
 import Dates
+import Printf
 
 import ..isvalid_barekey_char
 
+const is_printf_bugged = Printf.@sprintf("%.4i", 1) == "1"
+
+function print_toml_escaped(io::IO, s::AbstractString)
+    for c::AbstractChar in s
+        if !isvalid(c)
+            error("TOML print: invalid character $(repr(c)) encountered when printing string")
+        end
+        if c == '\b'
+            Base.print(io, '\\', 'b')
+        elseif c == '\t'
+            Base.print(io, '\\', 't')
+        elseif c == '\n'
+            Base.print(io, '\\', 'n')
+        elseif c == '\f'
+            Base.print(io, '\\', 'f')
+        elseif c == '\r'
+            Base.print(io, '\\', 'r')
+        elseif c == '"'
+            Base.print(io, '\\', '"')
+        elseif c == '\\'
+            Base.print(io, "\\", '\\')
+        elseif Base.iscntrl(c)
+            Base.print(io, "\\u")
+            @static if is_printf_bugged
+                Base.print(io, lpad(UInt32(c), 4, '0'))
+            else
+                Printf.@printf(io, "%.4i", UInt32(c))
+            end
```

Comment:
Should be hex?

```suggestion
            Base.print(io, string(UInt32(c), base=16, pad=4))
```

(Or use `Printf.@printf(io, "%.4x", UInt32(c))` if that's faster... both allocate a `StringVector` internally though, so I guess they're about the same)

Reply:
:+1: